charts/linkerd2-cni/templates/cni-plugin.yaml

```diff
@@ -241,7 +241,7 @@ spec:
               command:
               - /bin/sh
               - -c
-              - kill -15 1
+              - kill -15 1; sleep 15s
```

Comment:
what is the purpose of this sleep?

Reply:
I added in to stop the script from spamming the logs and processing `SIGTERM` twice. Without the change, here's how the script behaves. This happens typically when we have another conf file injected.

```
Removing Linkerd's configuration file: /host/etc/cni/net.d/01-linkerd-cni.conf
Created CNI config /host/etc/cni/net.d/10-calico.conflist
Detected change in /host/etc/cni/net.d/: DELETE 01-linkerd-cni.conf
Detected change in /host/etc/cni/net.d/: DELETE 10-calico.conflist
Detected change in /host/etc/cni/net.d/: CREATE 10-calico.conflist
SIGTERM received, simply exiting...
Sending SIGKILL to inotifywait
Removing linkerd-cni artifacts.
Removing linkerd-cni config from /host/etc/cni/net.d/10-calico.conflist
SIGTERM received, simply exiting...
Removing linkerd-cni artifacts.
Removing linkerd-cni config from /host/etc/cni/net.d/10-calico.conflist
Removing linkerd-cni kubeconfig: /host/etc/cni/net.d/ZZZ-linkerd-cni-kubeconfig
Removing linkerd-cni binary: /host/opt/cni/bin/linkerd-cni
Exiting.
```

I think we put in `kill -15 1` in case the kubelet won't send a SIGTERM. Looking at the [k8s docs](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks) on lifecycle hooks:

```
A call to the PreStop hook fails if the container is already in a terminated or completed state and the hook must complete before the TERM signal to stop the container can be sent. The Pod's termination grace period countdown begins before the PreStop hook is executed, so regardless of the outcome of the handler, the container will eventually terminate within the Pod's termination grace period. No parameters are passed to the handler.
```

The kubelet will send a TERM signal as soon as the hook completes, but the `kill` command returns instantly, so when deleting the pod (e.g when uninstalling) means we'll get two signals. Since the second TERM will be sent after the grace period (default 30s), I thought a 15 second sleep would give us enough time to clean-up. My concern here is that triggering clean-up before it has finished will result in clean-up race conditions? (just a hunch, and it spams the logs). I do think having the kill in the preStop hook is good, perhaps I can add a comment on the sleep?

Reply:
Yeah, a comment sounds like a good idea since this is pretty non-obvious.